### PR TITLE
feat(schematics): use @ngrx/schematics with nx enhancements

### LINKED ISSUE #174: Clean up NgRx schematic
We are already doing a good job at generating some NgRx boilerplate. We can do a lot more though. Some improvements we should consider (ordered highest-to-lowest priority):

* Add store-freeze
*  Use `@ngrx/schematics` to generate much of the scaffolding

### PR DESCRIPTION
**Pull request by ThomasBurleson**
* ngrx collection now extends `@ngrx/schematics`
* modify the nrwl `ngrx` schematic to use `@ngrx/schematics` to generate the initial source for actions, reducers, and effects
* modify the generated source to conform to Nx standards
* generate init and interface files
* remove template files for action, effect, and reducer scaffolding
* modify collection to extend `@ngrx/schematics` so the ngrx commands are also easily accessible
* inject dataloaded enums into action class
* configure nx workspace devDep on @ngrx/schematics version
* override ngrx effects.spec.ts with nx templated version

fixes #174.

### CODE REVIEW

**Comment by netwobble** [comment]
@ThomasBurleson How is this progressing?  I'm extremely excited to see these two projects align.  I'm constantly checking/making decisions on what is best path of implementation between the two projects.  With this feature it seams like they will naturally converge.  Please feel free to reach out if you would require any testing help.  

Main item will be the class vs interface (Actions) and how this will be used with data persistence.  Also supporting more structured (directories) just makes sense for enterprise. 

Cheers,
AJ

File: packages/schematics/package.json [hunk 1]
```diff
@@ -1,21 +1,42 @@
 {
   "name": "@nrwl/schematics",
-  "version": "0.0.1",
+  "version": "0.0.2",
```
Comment (vsavkin) [inline packages/schematics/package.json]: this is a placehoder.

Comment (vsavkin) [reply packages/schematics/package.json]: We replace it with the real version when building the package.

File: packages/schematics/package.json [hunk 2]
```diff
@@ -1,21 +1,42 @@
...
+  "peerDependencies": { },
   "author": "Victor Savkin",
   "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/nrwl/nx/issues"
+  },
+  "homepage": "https://github.com/nrwl/nx#readme",
   "schematics": "./src/collection.json",
   "dependencies": {
+    "@ngrx/schematics": "^5.2.0",
```
Comment (vsavkin) [inline packages/schematics/package.json]: we should replace 5.2.0 with a placeholder, provision the version in lib-versions, and replace it when creating a package.

File: packages/schematics/package.json [hunk 3]
```diff
@@ -1,21 +1,42 @@
...
+    "@schematics/angular": "^0.4.6",
```
Comment (vsavkin) [inline packages/schematics/package.json]: ping it to a particular version. `@schematics/angular` isn't super stable, so I'd like our users to have the version with test with.

File: packages/schematics/src/collection/ngrx/ngrx.spec.ts [hunk 4]
```diff
@@ -172,4 +179,135 @@ describe('ngrx', () => {
...
+    const hasFile = file => expect(tree.exists(file)).toBeTruthy();
+    const tree = schematicRunner.runSchematic(
+      'ngrx',
+      {
+        name: 'user',
+        module: appConfig.appModule
+      },
+      appTree
+    );
+    // tree.visit((path) => console.log(path));
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/ngrx.spec.ts]: remove this

File: packages/schematics/src/collection/ngrx/schema.d.ts [hunk 5]
```diff
@@ -1,4 +1,4 @@
-export interface Schema {
+export interface NgrxOptions {
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/schema.d.ts]: Could you change it back? We use the same convention in all our schematics. The Schema type represents the contents of schema.json.

File: packages/schematics/src/collection/ngrx/schema.json [hunk 6]
```diff
@@ -35,11 +40,11 @@
     "directory": {
```
Comment (vsavkin) [inline packages/schematics/src/collection/ngrx/schema.json]: directory appears twice

File: packages/schematics/src/utils/file.spec.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { addApp } from './file';
```
Comment (vsavkin) [inline packages/schematics/src/utils/file.spec.ts]: it looks like you have some duplication:

file.ts and fileutils.ts
file.spec.ts and fileutils.spec.ts

File: packages/schematics/src/utils/module-utils.ts [hunk 8]
```diff
@@ -0,0 +1,218 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {
```
Comment (vsavkin) [inline packages/schematics/src/utils/module-utils.ts]: Why is this separate from ast-utils? Most of the helpers there are module-related as well.

File: packages/schematics/src/utils/module-utils.ts [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+  const list = target ? target.members : undefined;
+
+  if (!target) {
+    throw new Error(`Cannot find enum '${enumName}'`);
+  }
+
+  return pairs.reduce((buffer, it) => {
+    const addComma = !(list.hasTrailingComma || list.length === 0);
+    const member = it.value ? `${it.name} = '${it.value}'` : it.name;
+    const memberExists = () => {
```
Comment (vsavkin) [inline packages/schematics/src/utils/module-utils.ts]: why is this a function? it looks like it's called once without any arguments.

File: packages/schematics/src/utils/strings.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
+/**
```
Comment (vsavkin) [inline packages/schematics/src/utils/strings.ts]: This file does similar things to fileUtils. Since you added this one, you should reconcile the two.

**Review by vsavkin — DISMISSED** [review]

**Comment by vsavkin** [comment]
The ngrx E2E fails. Please take a look.

**Comment by ThomasBurleson** [comment]
@netwobble  - Here is a doc that describes this PR's [schematics-generated files](https://gist.github.com/ThomasBurleson/2c722efe1661adc673b2c34460f7973d). We are working to have this PR finished and merged this week.

File: packages/schematics/src/utils/module-utils.ts (hunk 8 above)
Comment (ThomasBurleson) [reply packages/schematics/src/utils/module-utils.ts]: Merged.

File: packages/schematics/src/collection/ngrx/schema.json (hunk 6 above)
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/schema.json]: fixed.

File: packages/schematics/src/collection/ngrx/schema.d.ts (hunk 5 above)
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/schema.d.ts]: fixed.

File: packages/schematics/src/collection/ngrx/ngrx.spec.ts (hunk 4 above)
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/ngrx.spec.ts]: fixed.

File: packages/schematics/package.json (hunk 1 above)
Comment (ThomasBurleson) [reply packages/schematics/package.json]: reverted.

File: packages/schematics/package.json (hunk 2 above)
Comment (ThomasBurleson) [reply packages/schematics/package.json]: I will make a separate issue after this PR.

File: packages/schematics/package.json (hunk 3 above)
Comment (ThomasBurleson) [reply packages/schematics/package.json]: done.

File: packages/schematics/src/utils/module-utils.ts (hunk 9 above)
Comment (ThomasBurleson) [reply packages/schematics/src/utils/module-utils.ts]: reduction of name/value pairs to array of InsertChange changes.

**Comment by ThomasBurleson** [comment]
@FrozenPandaz, @vsavkin - ready for last review.

**Review by vsavkin — APPROVED** [review]

File: packages/schematics/src/collection/ngrx/rules/update-effects.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+      `;
+      const astConstructor = findNodes(source, ts.SyntaxKind.Constructor)[0];
+      const lastParameter = findNodes(
+        astConstructor,
+        ts.SyntaxKind.Parameter
+      ).pop();
+
+      return new InsertChange(componentPath, lastParameter.end, toInsert);
+    };
+    // const removeEffect$ = () => {
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/rules/update-effects.ts]: fixed.

File: packages/schematics/src/collection/ngrx/index.ts [hunk 12]
```diff
@@ -2,238 +2,110 @@ import {
...
-  });
+/**
+ * @ngrx/schematics generates files in:
+ *    `/apps/<ngrxFeatureName>/`
+ *
+ * For Nx monorepo, however, we need to move the files to either
+ *  a) apps/<appName>/src/app/<directory>, or
+ *  b) libs/<libName>/src/<directory>
+ */
+function moveToNxMonoTree(ngrxFeatureName, nxDir, directory): Rule {
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/index.ts]: done

File: packages/schematics/src/collection/ngrx/index.ts [hunk 13]
```diff
@@ -2,238 +2,110 @@ import {
...
-      )}.init`;
+/**
+ * Rule to generate the Nx 'ngrx' Collection
+ */
+export default function generateNgrxCollection(_options: Schema): Rule {
+  return wrapIntoFormat(() => {
+    const options = normalizeOptions(_options);
+    const context: RequestContext = {
+      featureName: options.name,
+      moduleDir: findModuleParent(options.module),
```
[the remark this answers is not in the document]
Comment (ThomasBurleson) [reply packages/schematics/src/collection/ngrx/index.ts]: will use `path.dirname()`

**Comment by ThomasBurleson** [comment]
Closing this pull request to use PR #386 
 * Schematics updated to delete ngrx reducer.spec and use nx reducer.spec
 * `<feature>.reducer.ts` fixed with comments and interfaces renames
 * `App.module.ts` imports fixed

**Pull request closed without merging**